libsmtutil/z3/wrapper.cpp

```diff
@@ -0,0 +1,3505 @@
+// This file is auto-generated from genwrapper.py
+#include <z3.h>
+#include <tuple>
+#include <Z3Loader.h>
+
+namespace {
+
+template<typename T>
+struct FunctionTrait;
+
+template<typename R, typename... Args>
+struct FunctionTrait<R(*)(Args...)> {
+  using ResultType = R;
+  template<unsigned N>
+  using ArgType = std::tuple_element_t<N, std::tuple<Args...>>;
+};
+
+template<auto F>
+using ResultType = typename FunctionTrait<decltype(F)>::ResultType;
+
+template<auto F, unsigned N>
+using ArgType = typename FunctionTrait<decltype(F)>::template ArgType<N>;
+
+}
+
+extern "C" {
+
+void Z3_API Z3_set_error_handler(Z3_context c, Z3_error_handler h)
+{
+    static auto sym = reinterpret_cast<decltype(&Z3_set_error_handler)>(Z3Loader::get().loadSymbol("Z3_set_error_handler"));
+    sym(c, h);
+}
+
+ResultType<&Z3_algebraic_is_value> Z3_API Z3_algebraic_is_value(ArgType<&Z3_algebraic_is_value, 0> _0, ArgType<&Z3_algebraic_is_value, 1> _1)
```

Comment:
By the way: IMHO these are just beautiful :-D...

Reply:
The mechansim is by the way fool-proof. If any of these entry points were to result in an incompatible signature: compile-time error in the ``reinterpret_cast`` (ambiguous overloads in ``decltype``). If any of them is missing: link-time error: undefined reference. As weird as it may look, I can't imagine what could go wrong with it :-).